From 3fcb3163f9850ccd60e36c7e809d865a361ae09f Mon Sep 17 00:00:00 2001
Message-Id: <3fcb3163f9850ccd60e36c7e809d865a361ae09f.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Jeff McGee <jeff.mcgee@intel.com>
Date: Mon, 9 Jun 2014 10:50:55 -0500
Subject: [PATCH 150/312] FOR_UPSTREAM [VPG]: drm/i915/bdw: RC6 debugfs
 disabling for BDW

This patch adds Broadwell support to:

Author: Jeff McGee <jeff.mcgee@intel.com>
Date:   Mon Dec 16 17:36:50 2013 -0600

    FOR_UPSTREAM [VPG]: drm/i915: Add RC6 debugfs disabling

Change-Id: I1a7e8624923007d474c319537d8d62afdc0ba857
For: VIZ-3345
Signed-off-by: Jeff McGee <jeff.mcgee@intel.com>
[torourke: rebased and resubmitted]
Signed-off-by: Tom O'Rourke <Tom.O'Rourke@intel.com>
---
 drivers/gpu/drm/i915/i915_debugfs.c |    8 ++------
 drivers/gpu/drm/i915/intel_pm.c     |   20 +++++++++-----------
 2 files changed, 11 insertions(+), 17 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_debugfs.c b/drivers/gpu/drm/i915/i915_debugfs.c
index 07e3e34..76a7ebc 100644
--- a/drivers/gpu/drm/i915/i915_debugfs.c
+++ b/drivers/gpu/drm/i915/i915_debugfs.c
@@ -4696,9 +4696,7 @@ static int i915_rc6_disable_get(void *data, u64 *val)
 	struct drm_device *dev = data;
 	struct drm_i915_private *dev_priv = dev->dev_private;
 
-	if ((INTEL_INFO(dev)->gen < 6) ||
-	     IS_VALLEYVIEW(dev) ||
-	     IS_BROADWELL(dev))
+	if ((INTEL_INFO(dev)->gen < 6) || IS_VALLEYVIEW(dev))
 		return -ENODEV;
 
 	flush_delayed_work(&dev_priv->rps.delayed_resume_work);
@@ -4714,9 +4712,7 @@ static int i915_rc6_disable_set(void *data, u64 val)
 	struct drm_i915_private *dev_priv = dev->dev_private;
 	int ret;
 
-	if ((INTEL_INFO(dev)->gen < 6) ||
-	     IS_VALLEYVIEW(dev) ||
-	     IS_BROADWELL(dev))
+	if ((INTEL_INFO(dev)->gen < 6) || IS_VALLEYVIEW(dev))
 		return -ENODEV;
 
 	flush_delayed_work(&dev_priv->rps.delayed_resume_work);
diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index 876c2ab..84c4867 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -3601,9 +3601,7 @@ void gen6_set_rc6_mode(struct drm_device *dev, bool disable)
 {
 	struct drm_i915_private *dev_priv = dev->dev_private;
 
-	if ((INTEL_INFO(dev)->gen < 6) ||
-	     IS_VALLEYVIEW(dev) ||
-	     IS_BROADWELL(dev)) {
+	if ((INTEL_INFO(dev)->gen < 6) || IS_VALLEYVIEW(dev)) {
 		DRM_DEBUG_DRIVER("RC6 disable not supported\n");
 		return;
 	}
@@ -3614,6 +3612,11 @@ void gen6_set_rc6_mode(struct drm_device *dev, bool disable)
 
 	if (disable)
 		I915_WRITE(GEN6_RC_CONTROL, 0);
+	else if (IS_BROADWELL(dev))
+		I915_WRITE(GEN6_RC_CONTROL,
+			   dev_priv->rps.rc6_mask |
+			   GEN7_RC_CTL_TO_MODE |
+			   GEN6_RC_CTL_HW_ENABLE);
 	else
 		I915_WRITE(GEN6_RC_CONTROL,
 			   dev_priv->rps.rc6_mask |
@@ -3894,14 +3897,9 @@ static void gen8_enable_rps(struct drm_device *dev)
 	if (intel_enable_rc6(dev) & INTEL_RC6_ENABLE)
 		rc6_mask = GEN6_RC_CTL_RC6_ENABLE;
 	intel_print_rc6_info(dev, rc6_mask);
-	if (IS_BROADWELL(dev))
-		I915_WRITE(GEN6_RC_CONTROL, GEN6_RC_CTL_HW_ENABLE |
-				GEN7_RC_CTL_TO_MODE |
-				rc6_mask);
-	else
-		I915_WRITE(GEN6_RC_CONTROL, GEN6_RC_CTL_HW_ENABLE |
-				GEN6_RC_CTL_EI_MODE(1) |
-				rc6_mask);
+
+	dev_priv->rps.rc6_mask = rc6_mask;
+	gen6_set_rc6_mode(dev, dev_priv->rps.rc6_disable);
 
 	/* 4 Program defaults and thresholds for RPS*/
 	I915_WRITE(GEN6_RPNSWREQ,
-- 
1.7.9.5

